src/iter.rs

```diff
@@ -0,0 +1,273 @@
+    /// ```
+    fn reading_frames(self) -> SmallVec<[Codons<Self::IntoIter>; 3]>
+    where
+        Self::IntoIter: Clone + ExactSizeIterator;
+
+    /// Returns iterator of reverse complement of contained nucleotides.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use quickdna::{Nucleotide, Nucleotides};
+    ///
+    /// use Nucleotide::*;
+    /// let dna = [C, G, A, T];
+    ///
+    /// assert!(dna.reverse_complement().eq([A, T, C, G]));
+    /// ```
+    fn reverse_complement(self) -> std::iter::Rev<Complement<Self::IntoIter>>
+    where
+        Self::IntoIter: DoubleEndedIterator;
+}
+
+impl<N, I, T> Nucleotides for T
+where
+    N: ToNucleotideLike,
+    I: Iterator<Item = N>,
+    T: IntoIterator<IntoIter = I>,
+{
+    fn codons(self) -> Codons<Self::IntoIter> {
+        Codons(self.into_iter())
+    }
+
+    fn complement(self) -> Complement<Self::IntoIter> {
+        Complement(self.into_iter())
+    }
+
+    fn reading_frames(self) -> SmallVec<[Codons<Self::IntoIter>; 3]>
+    where
+        Self::IntoIter: Clone + ExactSizeIterator,
+    {
+        let iter1 = self.into_iter();
+        let mut iter2 = iter1.clone();
+        iter2.next();
+        let mut iter3 = iter2.clone();
+        iter3.next();
+        let mut frames = SmallVec::from([iter1.codons(), iter2.codons(), iter3.codons()]);
+        frames.retain(|frame| frame.len() > 0);
+        frames
+    }
+
+    fn reverse_complement(self) -> std::iter::Rev<Complement<Self::IntoIter>>
+    where
+        Self::IntoIter: DoubleEndedIterator,
+    {
+        self.complement().rev()
+    }
+}
+
+/// Adapter yielding codons of the contained iterator.
+///
+/// This `struct` is created by the [`codons`](Nucleotides::codons)
+/// method on [`Nucleotides`]. See its documentation for more.
+#[derive(Clone, Debug)]
+pub struct Codons<I>(I);
+
+impl<N, I> Iterator for Codons<I>
+where
+    N: ToNucleotideLike,
+    I: Iterator<Item = N>,
+{
+    type Item = <N::NucleotideType as NucleotideLike>::Codon;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match (self.0.next(), self.0.next(), self.0.next()) {
+            (Some(n1), Some(n2), Some(n3)) => {
+                Some([n1, n2, n3].map(|n| n.to_nucleotide_like()).into())
+            }
+            _ => None,
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (min, max) = self.0.size_hint();
+        (min / 3, max.map(|m| m / 3))
+    }
+}
+
+impl<N, I> DoubleEndedIterator for Codons<I>
+where
+    N: ToNucleotideLike,
+    I: DoubleEndedIterator<Item = N>,
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        match (self.0.next_back(), self.0.next_back(), self.0.next_back()) {
+            (Some(n1), Some(n2), Some(n3)) => {
+                Some([n1, n2, n3].map(|n| n.to_nucleotide_like()).into())
+            }
+            _ => None,
+        }
+    }
+}
+
+impl<I> ExactSizeIterator for Codons<I>
+where
+    Self: Iterator,
+    I: ExactSizeIterator,
+{
```

Comment:
Providing an explicit implementation of `len` here that does `self.0.len() / 3` would be slightly more efficient than relying on the default impl, which does an `assert!` on the size_hint invariants.